static/js/main.js

```diff
@@ -0,0 +1,698 @@
+          this.files.push(uploaded_file);
+        }
+        show_messages(response.messages);
+      });
+
+      this.on("removedfile", function(file) {
+        // If this property is set we don't actually want to remove the file
+        // on the server, but just on the frontend
+        if (file.removed_on_fontend_only)
+          return;
+
+        // Cache this (dropzone) to access it in inner scope which has its
+        // own this
+        var thiz = this;
+        // Post the name of the functionary to remove
+        $.post({
+          method: "POST",
+          url: "/chaining/remove",
+          data: {"link_filename": file.name},
+          headers: _get_csrf_token_header(),
+          success: function(response) {
+            show_messages(response.messages);
+
+            // Re-add file (on client side) if server-side remove
+            // was not successful
+            if (response.error) {
+              thiz.emit("addedfile", file);
+              thiz.emit("complete", file, true);
+              thiz.files.push(file);
+            }
+          }
+        });
+      });
+    }
+  };
+  return new Dropzone($elem.get(0), opts);
+}
+
+/*
+ * Traverse `.ssc_steps` (c.f. software_supply_chain.html)
+ * and generate graph data suitable for `draw_graph`.
+ *
+ * Directed edges are created sequentially: E {node_i, node_i+1}
+ * Only "modifying" nodes have outdegree. A modifying node remains the edge
+ * source for all subsequent nodes until the next modifying node in the list.
+ *
+ * TODO: Find a way to display inspections
+ *
+ * Returns
+ * {
+ *   nodes: [
+ *     {
+ *       name: <step name>
+ *     }
+ *   ]
+ *   edges: [
+ *     {
+ *       source: <step name> ,
+ *       dest: <step name>
+ *     }, ...
+ *   ]
+ * }
+ */
+function generate_graph_from_ssc_steps() {
+  // Make a list of node objects (i.e.: [{name: <step_name>}] from
+  // step name input fields, ignore steps without name
+  var nodes = $.map($(".ssc-steps .ssc-step"),
+    function(elem) {
+      var step_name = $(elem).find("input[name='step_name[]']").val();
+      var step_modifies = ($(elem).find("input[name='step_modifies[]']").val()
+          === "true");
+      if (step_name) {
+        return {
+          name: step_name,
+          modifies: step_modifies
+        };
+      }
+      return;
+    });
+
+  var edges = [];
+  var src_i = null;
+  for (var i = 0; i < nodes.length - 1; i++) {
+    // Skip non-modifying steps for edge sources
+    // Only modifying steps, i.e. steps where materials and products are not
+    // equal, have an outdegree
+    if (nodes[i].modifies) {
+       src_i = i;
+    }
+    if (src_i !== null) {
+      edges.push({
+        source: nodes[src_i].name,
+        dest: nodes[i+1].name
+      })
+    }
+  }
+
+  return {
+    nodes: nodes,
+    edges: edges
+  };
+
+}
+
+/*
+ * Draw in-toto layout graph using dagre-d3 and D3.js
+ * https://github.com/cpettitt/dagre-d3/wiki
+ */
```

Comment:
I wonder if we want to store the d3 functions in a separate file.

Reply:
Definitely! I think the whole script can be dissected in many ways. There is already a related `FIXME` comment in the header block.  Do you think this is a blocker, or can we delegate the issue to a ticket?

Reply:
If it's already on a ticket, I think we can leave it as is.

Reply:
Will fix with https://github.com/in-toto/layout-web-tool/issues/31